staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_test.go

```diff
@@ -352,6 +356,125 @@ func TestCacherWatchSemanticInitialEventsExtended(t *testing.T) {
+		expectBookmarkOnWatchCache   bool
+		expectBookmarkOnEtcd         bool
+	}{
+		{
+			name:                       "common RPC > both get bookmarks",
+			separateCacheWatchRPC:      false,
+			expectBookmarkOnEtcd:       true,
+			expectBookmarkOnWatchCache: true,
+		},
+		{
+			name:                         "common RPC & watch cache context > both get bookmarks",
+			separateCacheWatchRPC:        false,
+			useWatchCacheContextMetadata: true,
+			expectBookmarkOnEtcd:         true,
+			expectBookmarkOnWatchCache:   true,
+		},
+		{
+			name:                       "separate RPC > only etcd gets bookmarks",
+			separateCacheWatchRPC:      true,
+			expectBookmarkOnEtcd:       true,
+			expectBookmarkOnWatchCache: false,
+		},
+		{
+			name:                         "separate RPC & watch cache context > only watch cache gets bookmarks",
+			separateCacheWatchRPC:        true,
+			useWatchCacheContextMetadata: true,
+			expectBookmarkOnEtcd:         false,
+			expectBookmarkOnWatchCache:   true,
+		},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SeparateCacheWatchRPC, tc.separateCacheWatchRPC)()
+			server, etcdStorage := newEtcdTestStorage(t, etcd3testing.PathPrefix())
+			_, cacher, server, terminate := testSetupCacher(t, server, etcdStorage)
+
+			if err := cacher.ready.wait(context.Background()); err != nil {
+				t.Fatalf("unexpected error waiting for the cache to be ready")
+			}
+			t.Cleanup(terminate)
+
+			requestProgressNotify := func() error {
+				ctx := context.Background()
+				if tc.useWatchCacheContextMetadata {
+					ctx = metadata.NewOutgoingContext(ctx, cacher.watchCache.waitingUntilFresh.contextMetadata)
+				}
+				return etcdStorage.RequestWatchProgress(ctx)
+			}
+
+			gotBookmark := watchCacheIsGettingBookmarks(t, context.Background(), server, cacher, requestProgressNotify)
+			if gotBookmark != tc.expectBookmarkOnWatchCache {
+				t.Errorf("Unexpected watch cache bookmark check result, got: %v, want: %v", gotBookmark, tc.expectBookmarkOnWatchCache)
+			}
+			gotBookmark = etcdWatchIsGettingBookmarks(context.Background(), etcdStorage, requestProgressNotify)
+			if gotBookmark != tc.expectBookmarkOnEtcd {
+				t.Errorf("Unexpected etcd bookmark check result, got: %v, want: %v", gotBookmark, tc.expectBookmarkOnEtcd)
+			}
+		})
+	}
+}
+
+func etcdWatchIsGettingBookmarks(ctx context.Context, etcdStorage storage.Interface, requestProgressNotify func() error) bool {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*2)
+	defer cancel()
+	opts := storage.ListOptions{ResourceVersion: "", Predicate: storage.Everything}
+	opts.Predicate.AllowWatchBookmarks = true
+	w, err := etcdStorage.Watch(ctx, "/pods/", opts)
+	if err != nil {
+		panic(err)
+	}
+	defer w.Stop()
+
+	gotBookmark := false
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for event := range w.ResultChan() {
```

Comment:
Why fo we need for loop? Can we get anything other than the expected bookmark?

Or you're effectively doing:
```
event, ok := <-w.ResultChan()
if !ok {
  return
}
```

Reply:
Wanted to explicitly wait for bookmark. We shouldn't get anything else, however I wanted to make sure the code properly reflects the intention.